numpy/core/tests/test_array_coercion.py

```diff
@@ -0,0 +1,524 @@
+        yield np.timedelta64(23, "s")
+        yield np.timedelta64("NaT", "s")
+
+        yield np.datetime64("NaT")  # generic units
+        yield np.datetime64("2020-06-07 12:43", "ms")
+
+        # Strings and unstructured void:
+        yield np.bytes_(b"1234")
+        yield np.unicode_("String")
+        yield np.void(b"4321")
+
+
+def is_parametric_dtype(dtype):
+    """Returns True if the the dtype is a parametric legacy dtype (itemsize
+    is 0, or a datetime without units)
+    """
+    if dtype.itemsize == 0:
+        return True
+    if issubclass(dtype.type, (np.datetime64, np.timedelta64)):
+        if dtype.name.endswith("64"):
+            # Generic time units
+            return True
+    return False
+
+
+class TestStringDiscovery:
+    @pytest.mark.parametrize("obj",
+            [object(), 1.2, 10**43, None, "string"])
+    def test_basic_stringlength(self, obj):
+        if not isinstance(obj, (str, int)):
+            pytest.xfail(
+                "The Single object (first assert) uses a different branch "
+                "and thus gives a different result (either wrong or longer"
+                "string than normally discovered).")
+
+        length = len(str(obj))
+        expected = np.dtype(f"S{length}")
+
+        assert np.array(obj, dtype="S").dtype == expected
+        assert np.array([obj], dtype="S").dtype == expected
+
+        # A nested array is also discovered correctly
+        arr = np.array(obj, dtype="O")
+        assert np.array(arr, dtype="S").dtype == expected
+
+    @pytest.mark.xfail(reason="Only single array unpacking is supported")
+    @pytest.mark.parametrize("obj",
+            [object(), 1.2, 10**43, None, "string"])
+    def test_nested_arrays_stringlength(self, obj):
+        length = len(str(obj))
+        expected = np.dtype(f"S{length}")
+        arr = np.array(obj, dtype="O")
+        assert np.array([arr, arr], dtype="S").dtype == expected
+
+    @pytest.mark.xfail(reason="Only single array unpacking is supported")
+    @pytest.mark.parametrize("arraylike", arraylikes())
+    def test_unpack_first_level(self, arraylike):
+        # We unpack exactly one level of array likes
+        obj = np.array([None])
+        obj[0] = np.array(1.2)
+        # the length of the included item, not of the float dtype
+        length = len(str(obj[0]))
+        expected = np.dtype(f"S{length}")
+
+        obj = arraylike(obj)
+        # casting to string usually calls str(obj)
+        arr = np.array([obj], dtype="S")
+        assert arr.shape == (1, 1)
+        assert arr.dtype == expected
+
+class TestScalarDiscovery:
+    def test_void_special_case(self):
+        # Void dtypes with structures discover tuples as elements
+        arr = np.array((1, 2, 3), dtype="i,i,i")
+        assert arr.shape == ()
+        arr = np.array([(1, 2, 3)], dtype="i,i,i")
+        assert arr.shape == (1,)
+
+    def test_char_special_case(self):
+        arr = np.array("string", dtype="c")
```

Comment:
is this operation the same as ```arr = np.array("string", dtype="S1")``` ?
if i use dtype as S1 it gives a different result:
```
>>> arr = np.array("string", dtype="S1")
>>> arr
array(b's', dtype='|S1')
```

Reply:
Yeah, its weird, `"c"` is the same as `"S1"`, but with the special case that `arr.dtype.char` is `c` and that here the result gives 6 individual characters instead of the `"string"`.  We should get rid of it some of these days IMO.

Reply:
IIRC, we have been planning to get rid of `c` for many years. It may be used somewhere, though. I think it was originally for data input for old applications.

Reply:
Yeah, we definitely should get rid of it, I just didn't feel like looking into it too deeply, because it seemed just as easy to add the special case for now.  I think `c` is actually partially deprecated on the C-API side, but not on the python side...